# Background information about the project


## Organism
Commensal Escherichia coli (sorbitol positive and negative isolates)

## Objective
Conduct comparative genomics to identify unique characteristics of commensal *E. coli* relative to Shiga-toxin producing E. coli (STEC) such as EDL933, RM6067, TW14588, FRIK1989 (used in FS19 study) that may help them outcompete or co-survive with STEC.

## Goals and Objectives from Project Plan NP 108 Food Safety (2021-2025):
* ###### Objective 2:
  * Commensal *E. coli* able to compete with STEC at the rectoanal junction will be isolated and E. coli encoded factors effective against STEC will be characterized... Collectively, the data will be used to generate targets for novel strategies against STEC attachment and colonization at the rectoanal junction.
* ###### Sub-objective 2.A:
  * Identify novel *E. coli* with encoded bacteriocins, or competitive nutritional networks that interfere with O157 viability, growth, or attachment to food animal intestinal mucosa.
* ###### Goal 2.A:
  * Identify commensal *E. coli* for use as competitor to O157 attachment, colonization, or survival.

## Literature related to project
### Project Plan NP 108 Food Safety (2021-2025): 108 Kudva 5030-32000-112-00D PrePlan.docx
  * #### Goal 2.A:
    * Identify commensal *E. coli* for use as competitor to O157 attachment, colonization, or survival.
  * #### Rationale 2.A:
    * While GIT colonization of reservoir animals by enteric foodborne pathogens is determined by many factors, including the ability to acquire needed nutrients in competition with other commensals. The ability to detect and utilize specific nutrients and host metabolites can give advantage to one organism over another <sup>13, 22, 23</sup> and competition between organisms in the same environment where resources may be scarce is an important factor in GIT colonization by enteric pathogens.<sup>14</sup> Despite O157 and commensal *E. coli* strains sharing the same core genome for central metabolic functions, O157 has acquired a large amount of additional DNA , some of which imparts it with the ability to utilize nutrients not utilized by many commensal *E. coli* strains. In addition, bacteria encode bacteriocins to eliminate competitive organisms in the same niche.<sup>186</sup> *E. coli* produced bacteriocins, are cytotoxic to susceptible similar or related bacterial strains but not to the producer strain.<sup>187, 188</sup> Thus, bacteriocin production may enable the producer strains to gain access to an established microbial community or niche, and on the same token prevent other strains to enter a niche occupied by a commensal strain.<sup>186</sup> With the advent of rapid and cost-effective whole genome sequencing technology, a large number of commensal *E. coli* from cattle, other ruminants, and livestock can be analyzed for the presence of genetic attributes enabling expression of specific carbon and nitrogen utilization networks also common to O157 strains. Also, genes encoding anti-O157 bacteriocins can be identified. The development of a probiotic *E. coli* to compete with O157 for the same nutritional sources and to produce anti-O157 bacteriocin could be a used to reducing O157 colonization in cattle.
  * #### Methods
    * ###### Whole genome sequencing and analysis:
      * Most *E. coli* can utilize many but not all of the following carbon/sugar substrates for their growth in the animal intestine: glucose, sucrose, galactose, arabinose, lactose, fucose, maltose, hexuornate, mannose, ribose, N-acetylglucosamine, N-acetylgalactosamine N-acetylneuraminate, sialic acid and D-gluconate.<sup>22, 23, 31, 189</sup> For example, O157, unlike most human commensal *E. coli*, can use galactose, hexuronate, mannose and ribose for in vivo growth. O157, like some other enteric pathogens and human commensal *E. coli* strains HS and Nissle 1917 carry a functional *eut* operon for utilization of ethanolamine (EA) as a nitrogen source.<sup>23, 27</sup> But how many of the above listed carbon/sugar substrates utilization pathways and in what combinations with or without *eut* genes bovine commensal *E. coli* use for growth and competitive advantage in cattle intestine is not known. Thus, we will use genomes of *E. coli* MG1655, *E. coli* HS, *E. coli* Nissle 1917, *E. coli* O157:H7 6564, and *E. coli* O157:H7 EDL933 as references (downloaded from NCBI Nucleotide Sequence database) for performing comparative genomics by using analytic tools described previously<sup>160</sup> in order to identify how many “structurally intact” carbon/sugars substrates utilization pathways and combinations thereof, along with genes for EA utilization, are present in the genomes of bovine commensal *E. coli* isolates.

## Carbon/sugars of interest
| Gene/Operon name | Description | Comments |
| -- | -- | -- |
| araBAD | arabinose | 10.1128/IAI.01386-07 |
| eut (17 genes including eutR transcriptional regulator) | ethanolamine utilization as nitrogen source | eut operon : https://academic.oup.com/jb/article-abstract/147/1/83/751773?redirectedFrom=fulltext, https://jb.asm.org/content/195/21/4947.long |
| fucAO<sup>1</sup>, fucK<sup>2</sup> | fucose | 10.1128/IAI.01386-07<sup>1</sup>, https://doi.org/10.1371/journal.pone.0053957<sup>2</sup> |
| galK | galactose | 10.1128/IAI.01386-07 |
| uxaB | galacturonate | 10.1128/IAI.01386-07 |
| gntK (idnK)<sup>1</sup>, edd/eda<sup>2</sup> | gluconate | 10.1128/IAI.01386-07<sup>1</sup>, https://iai.asm.org/content/64/9/3504<sup>2</sup> |
| manXYZ | glucosamine | 10.1128/IAI.01386-07 |
| uxaC | hexuronate<sup>1</sup>, glucuronate<sup>2</sup> | 10.1128/IAI.01386-07<sup>1</sup>, https://doi.org/10.1371/journal.pone.0053957<sup>2</sup> |
| lacZ | lactose | https://doi.org/10.1371/journal.pone.0053957 |
| manA | mannose | 10.1128/IAI.01386-07 |
| nanAT | N-acetylgalactosamine and galactosamine<sup>1</sup>, **N-acetylgalactosamine<sup>2</sup>** | 10.1128/IAI.01386-07<sup>1</sup>, https://doi.org/10.1371/journal.pone.0053957<sup>2</sup> |
| nagE | NAG (N-acetylglucosamine) | 10.1128/IAI.01386-07 |
| ?? | N-acetylneuraminic acid | 10.1128/IAI.01386-07, https://doi.org/10.1371/journal.pone.0053957 |
| rbsK | ribose | 10.1128/IAI.01386-07 |
| sacH | sucrose | https://doi.org/10.1371/journal.pone.0053957 |


## Virulence genes of interest
```
ler, escRSTU, sepZ, escD, espADB, tir, eae, cesT, espF, stcE, espC, hlyCABD, hlyE, pchABC, espP, efa1/lifA, toxB, stx1, and stx2
```

## Method papers and relevant sites
1. fastQC
  * Github: 10.1093/bioinformatics/btw354
  * https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
2. MultiQC DOI: 10.1093/bioinformatics/btw354
  * Github: https://github.com/ewels/MultiQC
  * https://multiqc.info/docs/
3. Mash DOI: 10.1186/s13059-016-0997-x
  * Github: https://github.com/marbl/Mash
  * http://mash.readthedocs.org
4. FastANI DOI: 10.1038/s41467-018-07641-9
  * Github: https://github.com/ParBLiSS/FastANI
5. Prokka DOI: 10.1093/bioinformatics/btu153
  * Github: https://github.com/tseemann/prokka
  * Prokka for loop DOI: https://doi.org/10.3389/fvets.2020.582297
6. Roary DOI: 10.1093/bioinformatics/btv421
  * Github: https://github.com/sanger-pathogens/Roary
  * Roary site: http://sanger-pathogens.github.io/Roary/
  * Great step-by-step tutorial: https://github.com/microgenomics/tutorials/blob/master/pangenome.md
7. PPanGGOLiN DOI: https://doi.org/10.1371/journal.pcbi.1007732
  * Github: https://github.com/labgem/PPanGGOLiN
8. RAxML DOI: 10.1093/bioinformatics/btu033
  * Github: https://github.com/stamatak/standard-RAxML
9. gifrop2
  * Github: https://github.com/Jtrachsel/gifrop

## Relevant email and other communication
List of genes from Vijay
```
You can search sequenced genomes of E. coli isolates for the genes/operons involved in assimilation of following carbon/sugar substrates :
glucose, sucrose, galactose, arabinose, lactose, fucose, maltose, hexuornate, mannose, ribose, N-acetylglucosamine, N-acetylgalactosamine N-acetylneuraminate, sialic acid and D-gluconate.  E. coli strains use these sugar substrates in various combinations/hierarchy for growth in the animal intestine:
E. coli O157, unlike most human commensal E. coli, can use galactose, hexuronate, mannose and ribose for in vivo growth.

In addition you can screen the sequenced genomes for:
Genes constituting eut operon (for ethanolamine utilization)
Genes encoding bacteriocins and microcins (especially microcins M and H47)
```

12Apr2021 Email correspondence with Vijay
```
Hi Kathy,
There are two genes in your list (ler, escRSTU, sepZ, escD, espADB, tagA, espC, bpf) I have changed to red font text. These two genes are tagA and bfp (corrected from bpf to bfp). I have added a few additional genes to your above list and deleted tagA and bfp as these are not found in enterohemorrhagic E. coli such as E. coli O157:H7. The modified list would be: ler, escRSTU, sepZ, escD, espADB, tir, eae, cesT, espF, stcE, espC, hlyCABD, hlyE, pchABC, espP, efa1/lifA, toxB, stx1, and stx2 for you to use in Blast search. The reason for making this change to gene list is follows:

tagA is a gene found in Vibrio species and it is a homologue of stcE gene (mucinase gene) of enterohemorrhagic E. coli (like E. coli O157:H7)
bfp is found on a plasmid in EPEC but this gene is not present in enterohemorrhagic E. coli such as  E. coli O157:H7
The pchABC genes are found in enterohemorrhagic E. coli (like E. coli O157:H7) and these are homologues of perC gene of enteropathogenic E. coli (EPEC). Both pchABC and perC genes activate LEE expression via activating ler
espP encodes an autotransported serine protease involved in the cleavage of human coagulation factor V
efa1/lifA encodes a lymphocyte inhibitory factor A
toxB encodes ToxB, a 362-kDa protein sharing amino acid sequence similarity with the large Clostridium toxin family, which is involved in adherence to epithelial cells in culture
stx1 and stx2 genes encode Shiga toxin 1 and Shiga toxin 2 (you may find these genes in some strains as standalone without other virulence genes)

Yes, EDL933 will be a good reference strain to use in your Blast search.

Hope this will be of help.
Thanks,
Vijay
```
