# Background information about the project


## Organism
Commensal Escherichia coli (sorbitol positive and negative isolates)

## Objective
Conduct comparative genomics to identify unique characteristics of commensal *E. coli* relative to Shiga-toxin producing E. coli (STEC) such as EDL933, RM6067, TW14588, FRIK1989 (used in FS19 study) that may help them outcompete or co-survive with STEC.

## Goals and Objectives from Project Plan NP 108 Food Safety (2021-2025):
* Taken from **108 Kudva 5030-32000-112-00D PrePlan.docx**
* ###### Objective 2:
  * Commensal *E. coli* able to compete with STEC at the rectoanal junction will be isolated and E. coli encoded factors effective against STEC will be characterized... Collectively, the data will be used to generate targets for novel strategies against STEC attachment and colonization at the rectoanal junction.
* ###### Sub-objective 2.A:
  * Identify novel *E. coli* with encoded bacteriocins, or competitive nutritional networks that interfere with O157 viability, growth, or attachment to food animal intestinal mucosa.
* ###### Goal 2.A:
  * Identify commensal *E. coli* for use as competitor to O157 attachment, colonization, or survival.

## Literature related to project
### Project Plan NP 108 Food Safety (2021-2025): 108 Kudva 5030-32000-112-00D PrePlan.docx
  * #### Goal 2.A:
    * Identify commensal *E. coli* for use as competitor to O157 attachment, colonization, or survival.
  * #### Experiment 2.A:
    * Culture commensal non-O157 *E. coli* strains inhabiting bovine GIT and using whole genome sequencing, screen isolates for the presence of specific metabolic utilization networks and/or bacteriocins for *in vitro* and *in vivo* co-culture competition assays.
  * #### Rationale 2.A:
    * While GIT colonization of reservoir animals by enteric foodborne pathogens is determined by many factors, including the ability to acquire needed nutrients in competition with other commensals. The ability to detect and utilize specific nutrients and host metabolites can give advantage to one organism over another <sup>13, 22, 23</sup> and competition between organisms in the same environment where resources may be scarce is an important factor in GIT colonization by enteric pathogens.<sup>14</sup> Despite O157 and commensal *E. coli* strains sharing the same core genome for central metabolic functions, O157 has acquired a large amount of additional DNA , some of which imparts it with the ability to utilize nutrients not utilized by many commensal *E. coli* strains. In addition, bacteria encode bacteriocins to eliminate competitive organisms in the same niche.<sup>186</sup> *E. coli* produced bacteriocins, are cytotoxic to susceptible similar or related bacterial strains but not to the producer strain.<sup>187, 188</sup> Thus, bacteriocin production may enable the producer strains to gain access to an established microbial community or niche, and on the same token prevent other strains to enter a niche occupied by a commensal strain.<sup>186</sup> With the advent of rapid and cost-effective whole genome sequencing technology, a large number of commensal *E. coli* from cattle, other ruminants, and livestock can be analyzed for the presence of genetic attributes enabling expression of specific carbon and nitrogen utilization networks also common to O157 strains. Also, genes encoding anti-O157 bacteriocins can be identified. The development of a probiotic *E. coli* to compete with O157 for the same nutritional sources and to produce anti-O157 bacteriocin could be a used to reducing O157 colonization in cattle.
  * #### Methods
    * ##### Isolation of presumptive commensal *E. coli* from bovine intestinal contents:  
      * Feces collected from at least 20 cattle across various ages and breeds in different locations will be suspended PBS, vortexed vigorously and bacteria streaked on a sorbitol-MacConkey agar (SMAC) plate (Sigma Aldrich) prepared according to manufacturer’s recommendation. Plates will be incubated at 37°C for 24 h and colonies that appear pink will be re-streaked on SMAC plates and incubated as above. Each presumptive *E. coli* isolate purified from the feces will be confirmed by MALDI-TOF system at USDA-NVSL.
    * ###### Whole genome sequencing and analysis:
      * *E. coli* genomic DNA will be isolated using the QIAamp DNA Mini Kit (Qiagen, Valencia, CA) with on columnRNAse treatment. DNA concentration will be determined using a Qubit (Qiagen) and sequenced at the ISU DNA Core Laboratory using Nextera Flex Library Kit (Illumina) for whole genome sequencing. Raw data will be analyzed for genome assembly and annotation as described previously.<sup>158</sup> Most *E. coli* can utilize many but not all of the following carbon/sugar substrates for their growth in the animal intestine: glucose, sucrose, galactose, arabinose, lactose, fucose, maltose, hexuornate, mannose, ribose, N-acetylglucosamine, N-acetylgalactosamine N-acetylneuraminate, sialic acid and D-gluconate.<sup>22, 23, 31, 189</sup> For example, O157, unlike most human commensal *E. coli*, can use galactose, hexuronate, mannose and ribose for in vivo growth. O157, like some other enteric pathogens and human commensal *E. coli* strains HS and Nissle 1917 carry a functional *eut* operon for utilization of ethanolamine (EA) as a nitrogen source.<sup>23, 27</sup> But how many of the above listed carbon/sugar substrates utilization pathways and in what combinations with or without *eut* genes bovine commensal *E. coli* use for growth and competitive advantage in cattle intestine is not known. Thus, we will use genomes of *E. coli* MG1655, *E. coli* HS, *E. coli* Nissle 1917, *E. coli* O157:H7 6564, and *E. coli* O157:H7 EDL933 as references (downloaded from NCBI Nucleotide Sequence database) for performing comparative genomics by using analytic tools described previously<sup>160</sup> in order to identify how many “structurally intact” carbon/sugars substrates utilization pathways and combinations thereof, along with genes for EA utilization, are present in the genomes of bovine commensal *E. coli* isolates.

### References (title and link to paper) from Objective 2.A Project Plan:
* 13. [Nutritional basis for colonization resistance by human commensal Escherichia coli strains HS and Nissle 1917 against E. coli O157:H7 in the mouse intestine.](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0053957), Maltby et al., 2013. DOI: https://doi.org/10.1371/journal.pone.0053957
  * EMBL files missing glucuronate, galacuronate; however, based on a paper referenced by Maltby et al., glucoronate, galacuronate, and hexuronate are one sugar
  * "We note that the hexuronates, glucuronate, and galacturonate are considered here to be one sugar because mutation of the hexuronate catabolism pathway eliminates growth on both sugars."
  *  **Table 4 sugar utilization in the intestine by E. coli strains MG1655, EDL933, Nissle 1917, and HS**
  * "The new finding reported here, that E. coli Nissle 1917 and HS prevent colonization by EDL933, supports the hypothesis that nutrient consumption by commensal E. coli can limit nutrient availability to pathogens, which in turn points to the potential of probiotics for preventing disease."
* 14. [Factors affecting the microecology of the gut.](https://link.springer.com/chapter/10.1007/978-94-011-2364-8_6) not available (book chapter)
* 22. [Precolonized human commensal Escherichia coli strains serve as a barrier to E. coli O157:H7 growth in the streptomycin-treated mouse intestine](https://iai.asm.org/content/77/7/2876.long), Leatham et al., 2009. DOI: 10.1128/iai.00059-09
  * colicin and microcin production: microcin M and H47
* 23. [Commensal and Pathogenic Escherichia coli Metabolism in the Gut](https://www.asmscience.org/content/journal/microbiolspec/10.1128/microbiolspec.MBP-0006-2014), Conway et al., 2015. DOI: 10.1128/microbiolspec.MBP-0006-2014
  * Entner-Doudoroff pathway (ED): catabolize gluconate, encoded by **edd-eda operon**
  * **Table 2** sugars utilized by *E. coli* strains
  * see highlighted sentences and phrases!
  * Chang et al., 2004. DOI: https://doi.org/10.1073/pnas.0307888101
    * "Freter's nutrient-niche theory postulates that numerous ecological niches within the intestine are defined by nutrient availability"
    * "The majority of intestinal bacteria require a fermentable carbohydrate for growth, and fermentation is assumed to be the mode of metabolism used by most species"
    * "Most carbohydrate in the colon is in the form of mucosal polysaccharides, which are degraded by a few anaerobes that dominate the intestinal biota (9-11). The monosaccharides released from mucin and other mucosal glycoproteins support the growth of many intestinal bacteria such as E. coli, which does not make polysaccharide-degrading enzymes"
    * "Because the population of individual species is proportional to the concentration of its preferred nutrient(s), the relatively low population of E. coli in the intestine indicates that the concentration of its preferred nutrient(s) is low"
    * "It is important to understand how gastrointestinal pathogens acquire the nutrients necessary to infect their hosts and initiate the disease process. This understanding of host-pathogen interactions should build on knowledge of model systems."
    * "Thus, we conducted a genome-wide search for metabolic pathways used by E. coli K-12 to colonize the mouse intestine. Here we present the gene expression profile of E. coli MG1655 grown in vitro on mucus as the sole source of carbon and energy, conditions that mimic nutrient availability in the mouse intestine
    * "Comparison of phases 1 and 2 revealed other subtleties of growth on mucus: genes involved in catabolism of gluconate, glucuronate, sialic acid, NAG, mannose, maltose, and ribose were induced at the early time point, whereas genes involved in degradation of ethanolamine and fucose, anaerobic respiration, and the TCA cycle were induced at the later time point". See **Table 1** for carbon compound catabolism: Phase 1 (ansB gntK hycI nanA rbsK tnaA uxaA uxaB uxaC uxuA), Phase 2 (ansB eutJ eutM eutN eutP eutQ eutT fucA fucO gabT tnaA)
    * "In summary, the gene expression profiles of mucus-grown E. coli MG1655 identified genes involved in catabolism of NAG, sialic acid, glucosamine, fucose, ribose, glucuronate, galacturonate, gluconate, and maltose."
    * Several figures and tables of wildtype and mutants (mutations in catabolic pathways) in nutrient preference, fitness for colonization in competition with wildtype parent,
    * " Cultivation without shaking results in partially anaerobic conditions, which most closely reflects the limited oxygen availability of mucosal tissues in the mouse intestine"
    * "Two sugars that E. coli K-12 cannot grow on because it lacks the pathway genes, galactosamine and N-acetylgalactosamine, can be used by other E. coli strains"
    * "Further studies are needed to address the hypothesis that preferential utilization of different nutrients and/or the ability to switch to alternative nutrition is the basis for cocolonization of different E. coli strains and that direct competition for the same nutrient(s) leads to decreased fitness of some strains."
    * "The results of such studies may explain how various commensal strains affect nutrient availability, in some cases providing an open niche for infecting E. coli pathogens and in other cases preventing infection."
    * "For example, it is known that certain E. coli strains can prevent Shigella flexneri (41) and Salmonella typhimurium (42) infection in mice."
    * "there is the opportunity to build a nutritional framework for understanding how commensal strains serve as the first line of defense against intestinal infections."
  * Miranda et al., 2004. DOI: 10.1128/IAI.72.3.1666-1676.2004
    * O157:H7 EDL933 strain compete with MG1655 K12 strain in streptomyin-fed mice for glycolytic and gluconeogenic substrates in mice intestine; protection of intestine with precolonization of MG1655 from EDL933.
  * Sweeney et al., 1996. https://iai.asm.org/content/64/9/3504/article-info
    * "product of the eda gene, KDPG aldolase, is used in the metabolism of glucuronate, galacturonate, and gluconate"
    * "These data suggest that when an E. coli strain is unable to utilize the Entner-Doudoroff pathway in the intestine, even if inhibition by gluconate-derived KDPG is blocked by an edd mutation, the strain is unable to grow as rapidly in the intestine as one that has that ability"
    * "in the presence of gluconate, the Entner- Doudoroff intermediate KDPG accumulates in E. coli eda mutants (Fig. 1), and it is well known that an intracellular accumulation of KDPG has growth-inhibitory properties"
    * "The inhibitory nature of gluconate on the eda mutants"
    * "Streptomycin treatment selectively reduces the facultative micro- flora; however, the anaerobic population in the large intestine remains largely intact, and large numbers of different species coexist"
  * Autieri et al., 2007. DOI: 10.1128/IAI.00822-07
    * "We are interested in identifying the nutrients that Escherichia coli uses to successfully colonize the mouse large intestine in the face of competition from an extensive microflora."
    * "The mucus layer itself is in a dynamic state, constantly being synthesized and secreted by specialized goblet cells and degraded to a large extent by the indigenous intestinal microbes"
    * "Streptomycin treatment alters the microecology of the cecum, decreasing the populations of facultative anaerobes and some strict anaerobes and creating a niche for organisms such as E. coli"
    * "the overall number of strict anaerobes in the cecal contents of streptomycin-treated and conventional mice is essentially identical (1 × 109 to 2 × 109 CFU/g of contents)... This feature makes the streptomycin-treated mouse model the one of choice to identify the nutrients E. coli uses to colonize the mouse intestine in the face of competition from an extensive microflora."
    * "In the present study, we report that while studying the role of l-fucose as a nutrient for E. coli MG1655 and E. coli Nissle 1917 in the mouse intestine, we found that accumulation of l-fuculose-1-phosphate, an intermediate in the degradation of l-fucose in E. coli, stimulates utilization of d-ribose. The present studies therefore raise the possibility that pool sizes of **metabolic intermediates play a role in the selection of nutrients used by E. coli MG1655 and E. coli Nissle 1917 to achieve maximum colonizing ability.**"
    * "it is well known that **intracellular accumulation of sugar phosphates is inhibitory to E. coli growth**, although the reasons remain unclear"
    * "it is possible that a variety of metabolic intermediates comprise a tier of regulation designed for E. coli and perhaps other members of the microflora to more efficiently utilize the available limiting multiple nutrients to maintain growth rates sufficient to avoid washout from the intestine."
  * Jones et al., 2008. DOI: 10.1128/IAI.00096-08
    * "Glycogen is the primary carbon and energy storage molecule for enteric bacteria, including E. coli"
    * "Glycogen is synthesized during times when carbon is abundant but other nutrients are limiting"
    * "The results show that catabolism of the disaccharide maltose provides a competitive advantage in vivo, whereas degradation of exogenous forms of the more complex glucose polymer, maltodextrin, does not. Furthermore, we found that the utilization of maltose in vivo gives the pathogen E. coli EDL933 a competitive advantage during the initiation stage of colonization when competing with commensal E. coli MG1655"
    * "In conclusion, we wish to emphasize that therapies designed to specifically reduce intestinal nutrient availability may be effective because organisms such as E. coli O157:H7 live a hunger or famine existence, as indicated by their requirement for glycogen stores, which places the pathogens in peril of starvation before infection can begin, if administration of prebiotics or probiotics depletes their preferred nutrients"
* 27. [Ethanolamine controls expression of genes encoding components involved in interkingdom signaling and virulence in enterohemorrhagic Escherichia coli O157:H7](https://mbio.asm.org/content/3/3/e00050-12.long)
  * Read next!
  * description
* 31. [Comparison of carbon nutrition for pathogenic and commensal Escherichia coli strains in the mouse intestine](https://iai.asm.org/content/76/3/1143.long), Fabich et al., 2008. DOI: 10.1128/IAI.01386-07
  * This study "compared the pathogenic strain EDL933 and the commensal E. coli strain MG1655 modes of metabolism in vitro"
  * "Freter’s nutrient-niche hypothesis states that for a species to be successful in the intestine, it must grow faster than the washout rate and better than all other competitors on one or on a small number of growth-limiting nutrients."
  * "Commensal E. coli cells are dispersed throughout the mucus layer overlaying the intestinal epithelium"
  * galacturonate = glucuronate = arabinose = ribose
  * catabolite-repressing sugars: NAG, gluconate and galactose
  * "Mucus contains 13 monosaccharides (arabinose, fucose, galactose, gluconate, glucuronate, galacturonate, mannose, glucosamine, NAG (N-acetyl-glucosamine), galactosamine, GalNAc (N-acetyl-galactosamine), NANA (N-acetyl-neuraminic acid), and ribose) that are made available to E. coli by host epithelial cell turnover and polysaccharide-degrading anaerobes"
  * catabolic pathways for sugars
  * "Thus, we conclude that the nutrient-niche hypothesis should be revised further to include the concept that metabolic flexibility provides a competitive advantage in the intestine, i.e., the catabolic capacity of an intestinal species is tuned to variations in nutrient availability. In this light, pathogens might be especially good at taking advantage of available nutrients that are not consumed by the normal microbiota."
  * "Growth of E. coli is limited to monosaccharides, disaccharides, one trisaccharide, and maltodextrins up to 20 glucose units in length... While there are several glycoside hydrolases carried by the E. coli EDL933 and E. coli MG1655 genomes, none of them are secreted enzymes that can hydrolyze complex polysaccharides; there are no significant differences between pathogenic and commensal E. coli strains in this regard (20). Phenotypically, E. coli cannot degrade the complex polysaccharides found in mucus (23). It is the dominant anaerobes that degrade polysaccharides in the intestine... Thus, as we have suggested previously, it appears that E. coli grows on simple sugars released by anaerobic digestion."
  * "Thus, we were intrigued to find that the relative impact of catabolic lesions on colonization did not strictly follow the in vitro preference for these sugars."
  * "The physiology underlying the differences between in vivo versus in vitro nutrient preference is not clear."    
* 158. [Complete Genome Sequences of Curli-Negative and Curli-Positive Isolates of Foodborne Escherichia coli O157:H7 Strain 86-24](https://mra.asm.org/content/4/6/e01323-16)
  * print
  * description of paper
* 160. [Comparative genomics reveals structural and functional features specific to the genome of a foodborne Escherichia coli O157:H7](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-5568-6)
  * print
  * description of paper
* 186. [Bacteriocins: evolution, ecology, and application](https://www.annualreviews.org/doi/10.1146/annurev.micro.56.012302.161024?url_ver=Z39.88-2003&rfr_id=ori%3Arid%3Acrossref.org&rfr_dat=cr_pub++0pubmed)
  * print
  * description of paper
* 187. [Microcins in Enterobacteriaceae: Peptide Antimicrobials in the Eco-Active Intestinal Chemosphere](https://www.frontiersin.org/articles/10.3389/fmicb.2019.02261/full)
  * print
  * description of paper
* 188. [Colicin biology.](https://mmbr.asm.org/content/71/1/158.long)
  * print
  * description of paper
* 189. [Sugar metabolism, an additional virulence factor in enterobacteria.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7129668/)
  * description of paper  

| Gene/Operon name | Description | Comments |
| -- | -- | -- |
| araBAD | arabinose | 10.1128/IAI.01386-07 |
| eut (17 genes including eutR transcriptional regulator) | ethanolamine utilization as nitrogen source | eut operon : https://academic.oup.com/jb/article-abstract/147/1/83/751773?redirectedFrom=fulltext, https://jb.asm.org/content/195/21/4947.long |
| fucAO<sup>1</sup>, fucK<sup>2</sup> | fucose | 10.1128/IAI.01386-07<sup>1</sup>, https://doi.org/10.1371/journal.pone.0053957<sup>2</sup> |
| galK | galactose | 10.1128/IAI.01386-07 |
| uxaB | galacturonate | 10.1128/IAI.01386-07 |
| gntK (idnK)<sup>1</sup>, edd/eda<sup>2</sup> | gluconate | 10.1128/IAI.01386-07<sup>1</sup>, https://iai.asm.org/content/64/9/3504<sup>2</sup> |
| manXYZ | glucosamine | 10.1128/IAI.01386-07 |
| uxaC | hexuronate<sup>1</sup>, glucuronate<sup>2</sup> | 10.1128/IAI.01386-07<sup>1</sup>, https://doi.org/10.1371/journal.pone.0053957<sup>2</sup> |
| lacZ | lactose | https://doi.org/10.1371/journal.pone.0053957 |
| manA | mannose | 10.1128/IAI.01386-07 |
| nanAT | N-acetylgalactosamine and galactosamine<sup>1</sup>, **N-acetylgalactosamine<sup>2</sup>** | 10.1128/IAI.01386-07<sup>1</sup>, https://doi.org/10.1371/journal.pone.0053957<sup>2</sup> |
| nagE | NAG (N-acetylglucosamine) | 10.1128/IAI.01386-07 |
| ?? | N-acetylneuraminic acid | 10.1128/IAI.01386-07, https://doi.org/10.1371/journal.pone.0053957 |
| rbsK | ribose | 10.1128/IAI.01386-07 |
| sacH | sucrose | https://doi.org/10.1371/journal.pone.0053957 |


## Method papers and relevant sites
1. fastQC
  * Github: 10.1093/bioinformatics/btw354
  * https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
2. MultiQC DOI: 10.1093/bioinformatics/btw354
  * Github: https://github.com/ewels/MultiQC
  * https://multiqc.info/docs/
3. Mash DOI: 10.1186/s13059-016-0997-x
  * Github: https://github.com/marbl/Mash
  * http://mash.readthedocs.org
4. FastANI DOI: 10.1038/s41467-018-07641-9
  * Github: https://github.com/ParBLiSS/FastANI
5. Prokka DOI: 10.1093/bioinformatics/btu153
  * Github: https://github.com/tseemann/prokka
  * Prokka for loop DOI: https://doi.org/10.3389/fvets.2020.582297
6. Roary DOI: 10.1093/bioinformatics/btv421
  * Github: https://github.com/sanger-pathogens/Roary
  * Roary site: http://sanger-pathogens.github.io/Roary/
  * Great step-by-step tutorial: https://github.com/microgenomics/tutorials/blob/master/pangenome.md
7. PPanGGOLiN DOI: https://doi.org/10.1371/journal.pcbi.1007732
  * Github: https://github.com/labgem/PPanGGOLiN
8. RAxML DOI: 10.1093/bioinformatics/btu033
  * Github: https://github.com/stamatak/standard-RAxML
9. gifrop2
  * Github: https://github.com/Jtrachsel/gifrop



## Additional papers to consider
1. Adhesins for non-STEC isolates on leafy green vegetables: https://link.springer.com/article/10.1007/s42770-020-00251-y#Sec2
2. Mash-based analysis and pangenome of 10K+ E. coli genomes: https://www.nature.com/articles/s42003-020-01626-5#code-availability
3. Genome-scale metabolic reconstructions of E. coli strains: https://doi.org/10.1073/pnas.1307797110
  * Comprehensive list of metabolic genes from 2013 (gene names may have changed since then)

## Relevant email and other communication
List of genes from Vijay
```
You can search sequenced genomes of E. coli isolates for the genes/operons involved in assimilation of following carbon/sugar substrates :
glucose, sucrose, galactose, arabinose, lactose, fucose, maltose, hexuornate, mannose, ribose, N-acetylglucosamine, N-acetylgalactosamine N-acetylneuraminate, sialic acid and D-gluconate.  E. coli strains use these sugar substrates in various combinations/hierarchy for growth in the animal intestine:
E. coli O157, unlike most human commensal E. coli, can use galactose, hexuronate, mannose and ribose for in vivo growth.

In addition you can screen the sequenced genomes for:
Genes constituting eut operon (for ethanolamine utilization)
Genes encoding bacteriocins and microcins (especially microcins M and H47)
```

## KEGG
| Organism | KEGG genome information| KEGG Genome| KEGG Pathway | Comments |
| -- | -- | -- | -- | -- |
| Ecoli_O157H7_EDL933 | https://www.genome.jp/kegg-bin/show_organism?org=T00044 | https://www.genome.jp/dbget-bin/www_bget?gn:T00044 | https://www.genome.jp/dbget-bin/get_linkdb?-t+pathway+gn:T00044 | |
| Ecoli_K12_MG1655 | https://www.genome.jp/kegg-bin/show_organism?org=T00007 | https://www.genome.jp/dbget-bin/www_bget?gn:T00007 | https://www.genome.jp/dbget-bin/get_linkdb?-t+pathway+gn:T00007 | |
| Ecoli_O9_HS (commensal) | https://www.genome.jp/kegg-bin/show_organism?org=ecx | https://www.genome.jp/dbget-bin/www_bget?gn:T00591 | https://www.genome.jp/dbget-bin/get_linkdb?-t+pathway+gn:T00591 | |
| Ecoli_Nissle1917 | NA | NA | NA | ENA: https://www.ebi.ac.uk/ena/browser/view/GCA_000714595.1 |
| Ecoli_O157H7_NADC_6564 | NA | NA | NA | |
| Ecoli_O157H7_TW14588 | NA | NA | NA |

How to get a list of genes associated with metabolism of each carbon/sugar substrate? KEGG pathway has diagram and can click on individual genes, but hwo to get whole list?

Or maybe try UniProt? Or gapseq? 
