# Background information about the project


## Organism
Commensal Escherichia coli (sorbitol positive and negative isolates)

## Objective
Conduct comparative genomics to identify unique characteristics of commensal *E. coli* relative to Shiga-toxin producing E. coli (STEC) such as EDL933, RM6067, TW14588, FRIK1989 (used in FS19 study) that may help them outcompete or co-survive with STEC.

## Goals and Objectives from Project Plan NP 108 Food Safety (2021-2025):
* Taken from **108 Kudva 5030-32000-112-00D PrePlan.docx**
* ###### Objective 2:
  * Commensal *E. coli* able to compete with STEC at the rectoanal junction will be isolated and E. coli encoded factors effective against STEC will be characterized... Collectively, the data will be used to generate targets for novel strategies against STEC attachment and colonization at the rectoanal junction.
* ###### Sub-objective 2.A:
  * Identify novel *E. coli* with encoded bacteriocins, or competitive nutritional networks that interfere with O157 viability, growth, or attachment to food animal intestinal mucosa.
* ###### Goal 2.A:
  * Identify commensal *E. coli* for use as competitor to O157 attachment, colonization, or survival.

## Literature related to project
* ##### Project Plan NP 108 Food Safety (2021-2025): 108 Kudva 5030-32000-112-00D PrePlan.docx
  * ###### Goal 2.A:
    * Identify commensal *E. coli* for use as competitor to O157 attachment, colonization, or survival.
  * ###### Experiment 2.A:
    * Culture commensal non-O157 *E. coli* strains inhabiting bovine GIT and using whole genome sequencing, screen isolates for the presence of specific metabolic utilization networks and/or bacteriocins for *in vitro* and *in vivo* co-culture competition assays.
  * ###### Rationale 2.A:
    * While GIT colonization of reservoir animals by enteric foodborne pathogens is determined by many factors, including the ability to acquire needed nutrients in competition with other commensals. The ability to detect and utilize specific nutrients and host metabolites can give advantage to one organism over another <sup>13, 22, 23</sup> and competition between organisms in the same environment where resources may be scarce is an important factor in GIT colonization by enteric pathogens.<sup>14</sup> Despite O157 and commensal *E. coli* strains sharing the same core genome for central metabolic functions, O157 has acquired a large amount of additional DNA , some of which imparts it with the ability to utilize nutrients not utilized by many commensal *E. coli* strains. In addition, bacteria encode bacteriocins to eliminate competitive organisms in the same niche.<sup>186</sup> *E. coli* produced bacteriocins, are cytotoxic to susceptible similar or related bacterial strains but not to the producer strain.<sup>187, 188</sup> Thus, bacteriocin production may enable the producer strains to gain access to an established microbial community or niche, and on the same token prevent other strains to enter a niche occupied by a commensal strain.<sup>186</sup> With the advent of rapid and cost-effective whole genome sequencing technology, a large number of commensal *E. coli* from cattle, other ruminants, and livestock can be analyzed for the presence of genetic attributes enabling expression of specific carbon and nitrogen utilization networks also common to O157 strains. Also, genes encoding anti-O157 bacteriocins can be identified. The development of a probiotic *E. coli* to compete with O157 for the same nutritional sources and to produce anti-O157 bacteriocin could be a used to reducing O157 colonization in cattle.
  * ###### Methods
    * ###### Isolation of presumptive commensal *E. coli* from bovine intestinal contents:  
      * Feces collected from at least 20 cattle across various ages and breeds in different locations will be suspended PBS, vortexed vigorously and bacteria streaked on a sorbitol-MacConkey agar (SMAC) plate (Sigma Aldrich) prepared according to manufacturer’s recommendation. Plates will be incubated at 37°C for 24 h and colonies that appear pink will be re-streaked on SMAC plates and incubated as above. Each presumptive *E. coli* isolate purified from the feces will be confirmed by MALDI-TOF system at USDA-NVSL.
    * ###### Whole genome sequencing and analysis:
      * *E. coli* genomic DNA will be isolated using the QIAamp DNA Mini Kit (Qiagen, Valencia, CA) with on columnRNAse treatment. DNA concentration will be determined using a Qubit (Qiagen) and sequenced at the ISU DNA Core Laboratory using Nextera Flex Library Kit (Illumina) for whole genome sequencing. Raw data will be analyzed for genome assembly and annotation as described previously.<sup>158</sup> Most *E. coli* can utilize many but not all of the following carbon/sugar substrates for their growth in the animal intestine: glucose, sucrose, galactose, arabinose, lactose, fucose, maltose, hexuornate, mannose, ribose, N-acetylglucosamine, N-acetylgalactosamine N-acetylneuraminate, sialic acid and D-gluconate.<sup>22, 23, 31, 189</sup> For example, O157, unlike most human commensal *E. coli*, can use galactose, hexuronate, mannose and ribose for in vivo growth. O157, like some other enteric pathogens and human commensal *E. coli* strains HS and Nissle 1917 carry a functional *eut* operon for utilization of ethanolamine (EA) as a nitrogen source.<sup>23, 27</sup> But how many of the above listed carbon/sugar substrates utilization pathways and in what combinations with or without *eut* genes bovine commensal *E. coli* use for growth and competitive advantage in cattle intestine is not known. Thus, we will use genomes of *E. coli* MG1655, *E. coli* HS, *E. coli* Nissle 1917, *E. coli* O157:H7 6564, and *E. coli* O157:H7 EDL933 as references (downloaded from NCBI Nucleotide Sequence database) for performing comparative genomics by using analytic tools described previously<sup>160</sup> in order to identify how many “structurally intact” carbon/sugars substrates utilization pathways and combinations thereof, along with genes for EA utilization, are present in the genomes of bovine commensal *E. coli* isolates.
* References (title and link to paper) from Objective 2.A Project Plan:
  * 13. [Nutritional basis for colonization resistance by human commensal Escherichia coli strains HS and Nissle 1917 against E. coli O157:H7 in the mouse intestine.](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0053957)
    * print
  * 14. [Factors affecting the microecology of the gut.](https://link.springer.com/chapter/10.1007/978-94-011-2364-8_6) not available (book chapter)
  * 22. [Precolonized human commensal Escherichia coli strains serve as a barrier to E. coli O157:H7 growth in the streptomycin-treated mouse intestine](https://iai.asm.org/content/77/7/2876.long)
  * 23. [Commensal and Pathogenic Escherichia coli Metabolism in the Gut](https://www.asmscience.org/content/journal/microbiolspec/10.1128/microbiolspec.MBP-0006-2014)
    * Nutritional basis of intestinal colonization, *E. coli* metabolism
  * 27. [Ethanolamine controls expression of genes encoding components involved in interkingdom signaling and virulence in enterohemorrhagic Escherichia coli O157:H7](https://mbio.asm.org/content/3/3/e00050-12.long)
    * Need to print
    * description
  * 31. [Comparison of carbon nutrition for pathogenic and commensal Escherichia coli strains in the mouse intestine](https://iai.asm.org/content/76/3/1143.long)
    * print
    * description of paper
  * 158. [Complete Genome Sequences of Curli-Negative and Curli-Positive Isolates of Foodborne Escherichia coli O157:H7 Strain 86-24](https://mra.asm.org/content/4/6/e01323-16)
    * print
    * description of paper
  * 160. [Comparative genomics reveals structural and functional features specific to the genome of a foodborne Escherichia coli O157:H7](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-5568-6)
    * print
    * description of paper
  * 186. [Bacteriocins: evolution, ecology, and application](https://www.annualreviews.org/doi/10.1146/annurev.micro.56.012302.161024?url_ver=Z39.88-2003&rfr_id=ori%3Arid%3Acrossref.org&rfr_dat=cr_pub++0pubmed)
    * print
    * description of paper
  * 187. [Microcins in Enterobacteriaceae: Peptide Antimicrobials in the Eco-Active Intestinal Chemosphere](https://www.frontiersin.org/articles/10.3389/fmicb.2019.02261/full)
    * print
    * description of paper
  * 188. [Colicin biology.](https://mmbr.asm.org/content/71/1/158.long)
    * print
    * description of paper
  * 189. [Sugar metabolism, an additional virulence factor in enterobacteria.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7129668/)
    * description of paper  

## Relevant email and other communication

```
email contents
```
