# Notes, Ideas and Thoughts

#Skeleton
### Method
* Summary:
* Began on:
* Completed on:
* Platform:

#To do
* You can change your default Slurm account using running slurm-account-selector.sh on the login node.
  * To see all your Slurm accounts at any time, use “sacctmgr -Pns show user format=account”
* get pheatmap to work: https://towardsdatascience.com/pheatmap-draws-pretty-heatmaps-483dab9a3cc, https://slowkow.com/notes/pheatmap-tutorial/
* run ppanggolin
* check out anvi'o: https://merenlab.org/2016/11/08/pangenomics-v2/
* look at raxml file
    * visualize tree via FigTree
* check out roary and gifrop output
  * downloaded `*.Rtab`, `*.csv`, `*.tab`
  * need to download `*.txt`
* stx1, stx2 genes: what are the gene names?
* read up on gapseq: https://doi.org/10.1101/2020.03.20.000737
  * https://github.com/jotech/gapseq
* update 02b_MethodsSummary.md (conda, prokka, roary, ppanggolin, raxml)
* reduce space used on Ceres: /project/fsepru/kmou/FS19C/lane1/**
  * *.fastq.gz sequence data for samples 1-96
  * tar these files once finish running assembly slurm scripts and get covstats.txt
  * /project/fsepru/kmou/FS19C/polished_genomes: gzip fasta files from polished_genomes_100X, mash_all/, polishedgenomesprokka/, referencegenomes/
* lit review: read https://mbio.asm.org/content/3/3/e00050-12.long

# Journal Target
* [Journal submission guidelines](link)
