# Notes, Ideas and Thoughts

# To do
## FS19C
* new approach:
    * grab all STEC genomes from NCBI?
    * run prokka and roary of STEC genomes and commensal genomes together
    * DRAM
    * do GO enrichment analysis to find which STEC have sugar utilization pathways of interest, and then which commensals also have these pathways
* old approach:
    * run prokka with Escherichia coli of commensals + STECs via gifrop
    * parallel blast to make sure virulence genes exist in the strains
      * custom database of virulence genes: track which source (EDL933, Sakai, etc.)
* run `DRAM.py distill` for remaining 213 commensal and STEC annotated genomes - need to compile all genomes' `annotations.tsv` into one `annotations.tsv`
* update `02_Methods` (`08_DRAM.md`), `03_Results.md` (gifrop, DRAM)
* Run isolates through Daniel's github tools
  * https://github.com/nielsend/O157LineageAssignment
    `./LSPA6Long.sh [output CSV File Name] [Fasta File(s)]`
  * https://github.com/nielsend/ABRicateSequenceExtraction
    * Not as sure how applicable the output from this will be... pull out specific genes of interest and search for polymorphisms, etc...
* lit review: read https://mbio.asm.org/content/3/3/e00050-12.long
* ppanggolin on STEC + commensals on gbk files generated by prokka or DRAM?


## Ceres
* You can change your default Slurm account using running slurm-account-selector.sh on the login node.
  * To see all your Slurm accounts at any time, use “sacctmgr -Pns show user format=account”

# Random stuff
1. [Journal submission guidelines](link)

<details><summary>ClIcK iF yOu DaRe</summary>
Boo!! Congratulations
</details>

# Skeleton
### Method
* Summary:
* Began on:
* Completed on:
* Platform:
