# 03_QC
* Summary: QC by checking sequence quality with FastQC, MultiQC. Estimate distances between commensal *E. coli*, STEC, non-STEC *E. coli*, and non-*E. coli* with Mash and FastANI. Visualize distance estimates with MDS on R.
* Platform: Ceres

## (1) FastQC
* Summary: Ran fastqc on FS19C sequence data to assess sequence quality of individual reads for each sample, and to use output for MultiQC
* Ceres: `/project/fsepru/kmou/FS19C/lane1/fastqc` or `/project/fsepru/kmou/FS19C/lane1/fastqc/Fastqc_Sample95_96`

1. Run the following `fastqc.slurm` script (aka fastqc.slurm) as one job for samples 1-94 from 1st sequencing run, and 2nd job for samples 95 and 96 from second sequencing run:

  <details><summary>fastqc.slurm script</summary>
  ```
  #!/bin/bash
  #SBATCH --job-name=fastqc                              # name of the job submitted
  #SBATCH -p short                                    # name of the queue you are submitting to
  #SBATCH -N 1                                            # number of nodes in this job
  #SBATCH -n 16                                           # number of cores/tasks in this job, you get all 20 cores with 2 threads per core with hyperthreading
  #SBATCH -t 12:00:00                                      # time allocated for this job hours:mins:seconds
  #SBATCH -o "stdout.%j.%N.%x"                               # standard out %j adds job number to outputfile name and %N adds the node name
  #SBATCH -e "stderr.%j.%N.%x"                               # optional but it prints our standard error
  #SBATCH --mem=32G   # memory
  #Enter commands here:
  module load fastqc
  fastqc -t 16 *.fastq.gz Â 
  mv *fastqc* ./fastqc/
  #End of file
  ```
  </details>

##### Files generated (for each isolate):
  * *fastqc.zip
  * *fastqc.html
  * `Fastqc_Sample95_96/`

## (2) MultiQC
* Summary: Run MultiQC on fastqc output of samples 1-94 only
* Ceres: `/project/fsepru/kmou/FS19C/lane1/fastqc/FS19_1-94_multiqc_data`

Activate and run `prokka_env` conda environment on samples 1-94:
  ```
  salloc
  module load miniconda
  source activate /project/fsepru/kmou/prokka_env
  multiqc *.fastqc.zip
  conda deactivate
  ```

##### Files generated:
 * FS19_1-94_multiqc_report.html
 * FS19_1-94_multiqc_data directory

## (3) FastANI
* Summary: run fastANI on FS19C sequence data by running in conda environment to estimate Average Nucleotide Identity (ANI) using alignment-free approximate sequence mapping. It calculates distance between 2 sequences. Include reference genomes to see how all sequences cluster relative to one another and if there are any outliers. Jules had mentioned fastANI is more accurate than Mash, but Mash is faster.
* Ceres: `/project/fsepru/kmou/FS19C/polished_genomes_100X`

1. Activate `prokka_env` conda environment which has fastANI already installed
  ```
  salloc
  module load miniconda
  source activate /project/fsepru/kmou/prokka_env
  ```

2. Include the following reference genomes in same directory as polished commensal *E. coli* fasta files for fastANI analysis (downloaded from NCBI):
  * E. coli str. K-12 substr. MG1655 aka https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3
    * Saved as Ecoli_K-12_MG1655.fasta
  * E. coli HS aka https://www.ncbi.nlm.nih.gov/nuccore/NC_009800.1
    * Saved as Ecoli_HS.fasta
  * E. coli Nissle 1917 aka https://www.ncbi.nlm.nih.gov/nuccore/CP007799.1
    * Saved as Ecoli_Nissle1917.fasta
  * E. coli O157:H7 str. NADC 6564 aka https://www.ncbi.nlm.nih.gov/nuccore/CP017251
    * https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6408774/#CR28
    * Saved as Ecoli_NADC6564.fasta
  * E. coli O157:H7 str. EDL933 aka https://www.ncbi.nlm.nih.gov/nuccore/CP008957.1
    * https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6408774/
    * Saved as Ecoli_O157H7_EDL933.fasta
  * E. coli O157:H7 str. TW14588 https://www.ncbi.nlm.nih.gov/assembly/GCF_000155125.1/
    * Saved as Ecoli_TW14588_GCF_000155125.1_ASM15512v1_genomic.fna.gz
  * * (Enterobacteriaceae) Salmonella enterica subsp. enterica serovar Typhimurium str. LT2 aka https://www.ncbi.nlm.nih.gov/assembly/GCF_000006945.2
    * Saved as Styphimurium_LT2.fna.gz
  * (Non-Enterobacteriaceae, same phylum) Campylobacter jejuni subsp. jejuni NCTC 11168 aka https://www.ncbi.nlm.nih.gov/assembly/GCF_000009085.1
    * Saved as Cjejuni_11168.fna.gz
  * (Non-Proteobacteria, a Firmicutes) Clostridium saccharoperbutylacetonicum N1-4(HMT) aka https://www.ncbi.nlm.nih.gov/assembly/GCF_000340885.1
    * Saved as Clostridium_N1-4.fna.gz

2. Add all file names of 95 isolates and reference genomes in a `querylist2.txt` text file and run fastani with this file listed as query and as reference lists.
  ```
  ls -dv "$PWD"/* > quertylist2.txt
  fastANI --ql querylist2.txt --rl querylist2.txt -o fs19cfastanioutput2.out
  ```

##### Files generated:
  * FS19CfastANIoutput2.xlsx
  * fs19cfastanioutput2.out

## (4) Mash
* Summary: run Mash on FS19C sequence data by running in conda environment to compare results with fastANI. The *sketch* function converts a sequence or collection of sequences into a MinHash sketch. The *dist* function compares two sketches and returns an estimate of the Jaccard index, *P* value, and Mash distance (estimates rate of sequence mutation under a simple evolutionary model). Also include reference genomes to see how all sequences cluster relative to one another and if there are any outliers. Jules had mentioned fastANI is more accurate than Mash, but Mash is faster.
* Ceres: `/project/fsepru/kmou/FS19C/polished_genomes_100X`

1. Activate `prokka_env` conda environment. Mash is already installed in this environment.
  ```
  salloc
  module load miniconda
  source activate /project/fsepru/kmou/prokka_env
  ```

2. Make a new directory `mash_all` that has all polished genomes and reference genomes together. Reference genomes include:
* E. coli str. K-12 substr. MG1655 aka https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3
  * Saved as Ecoli_K-12_MG1655.fasta
* E. coli HS aka https://www.ncbi.nlm.nih.gov/nuccore/NC_009800.1
  * Saved as Ecoli_HS.fasta
* E. coli Nissle 1917 aka https://www.ncbi.nlm.nih.gov/nuccore/CP007799.1
  * Saved as Ecoli_Nissle1917.fasta
* E. coli O157:H7 str. NADC 6564 aka https://www.ncbi.nlm.nih.gov/nuccore/CP017251
  * https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6408774/#CR28
  * Saved as Ecoli_NADC6564.fasta
* E. coli O157:H7 str. EDL933 aka https://www.ncbi.nlm.nih.gov/nuccore/CP008957.1
  * https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6408774/
  * Saved as Ecoli_O157H7_EDL933.fasta
* E. coli O157:H7 str. TW14588 https://www.ncbi.nlm.nih.gov/assembly/GCF_000155125.1/
  * Saved as Ecoli_TW14588_GCF_000155125.1_ASM15512v1_genomic.fna.gz
* * (Enterobacteriaceae) Salmonella enterica subsp. enterica serovar Typhimurium str. LT2 aka https://www.ncbi.nlm.nih.gov/assembly/GCF_000006945.2
  * Saved as Styphimurium_LT2.fna.gz
* (Non-Enterobacteriaceae, same phylum) Campylobacter jejuni subsp. jejuni NCTC 11168 aka https://www.ncbi.nlm.nih.gov/assembly/GCF_000009085.1
  * Saved as Cjejuni_11168.fna.gz
* (Non-Proteobacteria, a Firmicutes) Clostridium saccharoperbutylacetonicum N1-4(HMT) aka https://www.ncbi.nlm.nih.gov/assembly/GCF_000340885.1
  * Saved as Clostridium_N1-4.fna.gz

3. Sketch all the genomes together.
    ```
    mash sketch -p 1 -o /project/fsepru/kmou/FS19C/polished_genomes_100X/mash_all/ *.fasta
    ```
4. Run distance estimates comparing sketch genomes with itself. This runs really fast (less than a second).
  ```
  mash dist -p 1 .msh .msh > distances.tab
  ```
  * `distances_thirdrun.tab` columns: reference_id, query_id, mash_distance, pvalue, matching_hashes

##### Files generated:
  * distances_thirdrun.tab
  * FS19Cmashdistances_thirdrun.xlsx

## (5) Visualize ANI and Mash distance calculations with MDS and heatmap
* Summary: Made distance matrix from mash and fastani output to create heatmap and MDS to visualize clustering and identify any outliers. The MDS was a bit hard to decipher what was an outlier, so I ran a heatmap to see how fastANI and mash compared and whether the pairwise comparisons were similar between the two, including heatmap of pearson correlation coefficients.
* Platform: R Studio

1. Run `qc_mds.R`.

<details><summary>qc_mds.R script</summary>

  ```
  #####################################################################################################
  #FS19C QC - MDS of fastANI and Mash output
  #Kathy Mou

    #Purpose: Convert ANI pairwise genome-genome similarity calculations of fastANI and mash output to a distance matrix and visualize distance matrices with MDS and heatmaps.

    #Load library packages
  library(ggplot2)
  library(tidyverse)
  library(ggrepel) #for geom_text_repel()
  library(pheatmap)
  library(gtools) #for mixedorder() to sort reference_id numerically, which has mixed numeric and characters

    sessionInfo()
  #R version 4.0.2 (2020-06-22)

    #####################################################################################################

    #In mash and fastani files, I removed extra path names (/home/..., _pol.fasta and .fasta) from elements in reference_id and query_id
  #in excel before importing distances.tab to R so that distance calculation would work
  #Otherwise, it produces errors saying duplicate 'row.names' are not allowed, non-unique values when setting 'row.names'
  #fastANI gives you ANI similarity in percentage, mash already did the distance calculation for you (hence the small decimal numbers)
  #so you don't need to do distance calculation (1-ANI).
  #You will need to convert fastANI ANI percentage values to decimal and then subtract that from 1 to get distance calculation.
  #I did a little experiment to see if reordering reference_id of mash file (distance.tab) with mixedorder() would change MDS and
  #heatmap structure compared to if I didn't reorder reference_id. In conclusion, no difference, so don't need to reorder reference_id

    ################################### FastANI ##################################################################
  ######## Import fastani file, remove orthologous_matches, total_seq_fragments columns ########
  fast_tab <- read_tsv('./Files/fs19cfastanioutput2.out.tab', col_names = c("reference_id", "query_id", "ani", "orthologous_matches", "total_seq_fragments")) %>%
  select(-orthologous_matches) %>%
  select(-total_seq_fragments) %>%
  mutate(ani_dist= 1-(ani/100)) %>% #add new ani_dist column to calculate distance
  select(-ani)

    ######## Make distance matrix ########
  fast_dist <- fast_tab %>%
  pivot_wider(names_from = reference_id, values_from=ani_dist) %>%
  column_to_rownames(var='query_id')

    ######## Heatmap of fastani distance matrix as another way of looking at clustering ########
  #Adapted code from: https://www.datanovia.com/en/blog/clustering-using-correlation-as-distance-measures-in-r/
  fast_heatmap <- pheatmap(fast_dist, scale = "row", main = "fastANI distance matrix heatmap")
  fast_heatmap
  ggsave("FS19C_fastani_heatmap.tiff", plot=fast_heatmap, width = 13, height = 14, dpi = 500, units =c("in"))

    ######## Pairwise correlation between samples ########
  #by columns
  cols.cor <- cor(fast_dist, use = "pairwise.complete.obs", method='pearson')
  #by rows
  rows.cor <- cor(t(fast_dist), use = 'pairwise.complete.obs', method='pearson')

    ######## Plot heatmap of pairwise correlations ########
  fast_corr_heatmap <- pheatmap(fast_dist, scale = 'row',
                              clustering_distance_cols = as.dist(1 - cols.cor),
                              clustering_distance_rows = as.dist(1 - rows.cor),
                              main = "fastANI pairwise correlation heatmap")
  fast_corr_heatmap
  ggsave("FS19C_fastani_correlation_heatmap.tiff", plot=fast_corr_heatmap, width = 11, height = 14, dpi = 500, units =c("in"))

    ######## Calculate distance for MDS ########
  fast_dist2 <- as.dist(fast_dist) #distance matrix computation that computes distances between rows of a data matrix
  head(fast_dist2)

    ######## Generate MDS ########
  fast_mds <- cmdscale(fast_dist2) %>% as.data.frame() %>%
  rownames_to_column(var='reference_id')
  #cmdscale = classic MDS of a data matrix

    ######## Plot MDS ########
  plot_fast_mds <-
  fast_mds %>% ggplot(aes(x=V1, y=V2)) +
  geom_point()+
  geom_text_repel(aes(label=reference_id), max.overlaps = 50) +
  labs(x='MDS1', y='MDS2') +
  ggtitle("fastANI MDS")
  plot_fast_mds
  ggsave("FS19C_fastaniMDS.tiff", plot=plot_fast_mds, width = 9, height = 8, dpi = 500, units =c("in"))

    ############################################ Mash #########################################################
  ######## Import mash file, add ani_dist column, remove pvalue, matching_hashes columns ########
  mash_tab <- read_tsv('./Files/distances_thirdrun.tab', col_names = c("reference_id", "query_id", "ani_dist", "pvalue", "matching_hashes")) %>%
  select(-pvalue) %>%
  select(-matching_hashes)

    ######## Make distance matrix ########
  #mash_dist with mash_tab
  mash_dist <- mash_tab %>%
  arrange(reference_id) %>%
  pivot_wider(names_from = reference_id, values_from=ani_dist) %>%
  column_to_rownames(var='query_id')

    ######## Heatmap of mash distance matrix as another way of looking at clustering ########
  #using mash_dist
  mash_heatmap <- pheatmap(mash_dist, scale = "row", main = "mash distance matrix heatmap")
  mash_heatmap
  ggsave("FS19C_mash_heatmap.tiff", plot=mash_heatmap, width = 13, height = 14, dpi = 500, units =c("in"))

    ######## Pairwise correlation between samples (columns) ########
  #using mash_dist
  cols.cor.mash <- cor(mash_dist, use = "pairwise.complete.obs", method='pearson')
  #pairwise correlation between samples (rows)
  rows.cor.mash <- cor(t(mash_dist), use = 'pairwise.complete.obs', method='pearson')

    ######## Plot heatmap of pairwise correlations ########
  #using mash_dist
  mash_corr_heatmap <- pheatmap(mash_dist, scale = 'row',
                              clustering_distance_cols = as.dist(1 - cols.cor.mash),
                              clustering_distance_rows = as.dist(1 - rows.cor.mash),
                              main = "mash pairwise correlation heatmap")
  mash_corr_heatmap
  ggsave("FS19C_mash_correlation_heatmap.tiff", plot=mash_corr_heatmap, width = 11, height = 14, dpi = 500, units =c("in"))

    ######## Calculate distance for MDS ########
  #using mash_dist
  mash_distA <- as.dist(mash_dist) #distance matrix computation that computes distances between rows of a data matrix
  head(mash_distA)

    ######## Generate MDS ########
  #using mash_distA
  mash_mdsA <- cmdscale(mash_distA) %>% as.data.frame() %>%
  rownames_to_column(var='reference_id')
  #cmdscale = classic MDS of a data matrix

    ######## Plot MDS ########
  #using mash_mdsA
  plot_mash_mdsA <-
  mash_mdsA %>% ggplot(aes(x=V1, y=V2)) +
  geom_point()+
  geom_text_repel(aes(label=reference_id), max.overlaps = 50) +
  labs(x='MDS1', y='MDS2') +
  ggtitle("Mash MDS")
  plot_mash_mdsA
  ggsave("FS19C_mashMDS.tiff", plot=plot_mash_mdsA, width = 9, height = 8, dpi = 500, units =c("in"))
  ```
  </details>

2. How to determine outliers in MDS plot:
  * https://doi.org/10.1093/bioinformatics/btz964: They use heatmaps of distance matrices to type bacteria. You can try generating a heatmap to see how it compares between fastANI and mash.
  * Followed this site for generating heatmap from distance matrix and correlation: https://www.datanovia.com/en/blog/clustering-using-correlation-as-distance-measures-in-r/
